fix(ruby): properly highlight %r{} regex literals containing nested patterns

j-erasmus opened the pull request:
### Summary

This fixes an issue where `%r{}`-style regular expressions in Ruby were not correctly highlighted when they contained nested patterns or complex escape sequences. Patterns like `(\.{2}|\A/)` would either break highlighting or be ignored entirely due to improper handling of nesting inside `%r{}`.

Resolves: https://github.com/highlightjs/highlight.js/issues/4281

---

### Root Cause

Highlight.js previously treated `%r{}` regex literals as flat strings. This caused the parser to:
- Fail on nested constructs like `{2}` (quantifiers)
- Break on position anchors like `\A`
- Terminate the regex early or skip highlighting entirely

---

### Fix

- Introduces a top-level `REGEXP` mode to match `%r{}`, `%r()`, `%r[]`, `%r!`, etc. anywhere in Ruby code
- Enables nested brace parsing inside `%r{}` using `contains: ['self']`
- Removes `%r` variants from the legacy `RE_STARTERS_RE` block (which still handles `/.../` contexts)
- Keeps other regex behavior intact and scoped

---

### Screenshots

- [sample file](https://gitlab.com/gitlab-org/gitlab/-/blob/3680e8318044d7d2ee7e7c4e42fa5588001c9008/app/models/project.rb)

| Before  | After |
| ------------- | ------------- |
|  ![Screenshot 2025-06-27 at 10 39 06](https://github.com/user-attachments/assets/b464660c-2d5b-47c7-8df5-d5bd72a2e1dc)  |   ![Screenshot 2025-06-27 at 10 39 24](https://github.com/user-attachments/assets/3aa94a43-f69d-47cf-a7eb-d18b6e092df0) |

j-erasmus committed:
fix(ruby): properly highlight nested %r{} regex

Previously, the Ruby grammar only recognized %r{} regex literals
in limited contexts, causing patterns like %r{(\.{2}|\A/)} to
break syntax highlighting.

This fix:
- Adds a top-level REGEXP mode to support all %r{} variations globally
- Enables nested brace handling inside %r{...}
- Removes redundant %r handling tied to RE_STARTERS_RE
- Adds a test case to regexes.txt to prevent regression

joshgoebel commented on src/languages/ruby.js:
```diff
@@ -227,6 +227,27 @@ export default function(hljs) {
+    illegal: /\n/,
+    variants: [
+      { begin: /%r\{/, end: /\}[a-z]*/ },
+      { begin: '%r\\(', end: '\\)[a-z]*' },
+      { begin: '%r!', end: '![a-z]*' },
+      { begin: '%r\\[', end: '\\][a-z]*' },
+      { begin: '/', end: '/[a-z]*' }
```
All of these should be regex literals, not strings.

joshgoebel commented on src/languages/ruby.js:
```diff
@@ -227,6 +227,27 @@ export default function(hljs) {
+      hljs.BACKSLASH_ESCAPE,
+      SUBST,
+      {
+        begin: /\{/, end: /\}/,
```
Maybe just handling `[]` with escapes inside (to allow escaping a other closing `]`) would be sufficient though?

j-erasmus replied:
@joshgoebel this should now be fixed, I also added some tests to verify these cases.

j-erasmus committed:
fix(ruby): handle braces correctly inside regex

- Add character class handling to prevent {} inside [] from being treated as interpolation
- Support escaped closing brackets (\]) within character classes
- Ensure quantifiers like {3} still work outside character classes
- Add test cases to verify edge cases are handled properly

Yang062 commented:
101142167176

***@***.***

Jacques Erasmus ***@***.***> 于 2025年7月4日周五 19:49写道：

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/languages/ruby.js
> <https://github.com/highlightjs/highlight.js/pull/4286#discussion_r2185158611>
> :
>
> > @@ -227,6 +227,27 @@ export default function(hljs) {
>      ]
>    };
>
> +  const REGEXP = {
> +    className: 'regexp',
> +    contains: [
> +      hljs.BACKSLASH_ESCAPE,
> +      SUBST,
> +      {
> +        begin: /\{/, end: /\}/,
>
> @joshgoebel <https://github.com/joshgoebel> this should now be fixed, I
> also added some tests to verify these cases.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/highlightjs/highlight.js/pull/4286#discussion_r2185158611>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BK2N76VR5FECLYCE47MX7X33GZS4NAVCNFSM6AAAAACAIM4ROGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDSOBWHA3DINRWHA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

j-erasmus committed:
Merge branch 'main' into 466023-fix-ruby-syntax-highlighting

j-erasmus commented:
@joshgoebel any updates here? I've addressed the comments you left previously. Please take a look/merge when you get a chance 🙇

j-erasmus commented:
@joshgoebel just a friendly ping here. Do you think we can get this merged? Alternatively we can patch it in our project.
